File: Tries1.java
//Time complexity - O(M) - M is the key length
//Space complexity - O(M)


class Trie {
    
    TrieNode root;
    
    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
    }

    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode curr = root;
        for(char c: word.toCharArray()){
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        curr.isEnd = true;
    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode curr = root;
        for(char c: word.toCharArray()){
            if(!curr.children.containsKey(c)) return false;
            curr = curr.children.get(c);
        }
        return curr.isEnd;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for(char c: prefix.toCharArray()){
            if(!curr.children.containsKey(c)) return false;
            curr = curr.children.get(c);
        }
        return true;
    }
}

File: Tries-1.py
# TC/SC: O(M)/O(M)
class Trie:

    def __init__(self):
        self.trieStorage = {}
        self.endCharacter = '*'
        

    def insert(self, word: str) -> None:
        node = self.trieStorage
        for i in word:
            if i not in node:
                node[i] = {}
            node = node[i]
        
        node[self.endCharacter] = True
                
        
    def search(self, word: str) -> bool:
        node = self.trieStorage
        for i in word:
            if i not in node:
                return False
            node = node[i]
        
        return self.endCharacter in node
        

    def startsWith(self, word: str) -> bool:
        node = self.trieStorage
        for i in word:
            if i not in node:
                return False
            node = node[i]
        
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
